google-cloud-storage/src/main/java/com/google/cloud/storage/SignatureInfo.java

```diff
@@ -0,0 +1,177 @@
+    
+    if (canonicalizedExtensionHeaders != null) {
+      
+      List<String> orderedKeys = new ArrayList<>(canonicalizedExtensionHeaders.keySet());
+      Collections.sort(orderedKeys);
+      
+      for (String key : orderedKeys) {
+        payload.append(key).append(HEADER_SEPARATOR).append(canonicalizedExtensionHeaders.get(key))
+        .append(COMPONENT_SEPARATOR);
+      }
+    }
+    payload.append(canonicalizedResource);
+
+    return payload.toString();
+  }
+
+  public HttpMethod getHttpVerb() {
+    return httpVerb;
+  }
+
+  public String getContentMd5() {
+    return contentMd5;
+  }
+
+  public String getContentType() {
+    return contentType;
+  }
+
+  public long getExpiration() {
+    return expiration;
+  }
+
+  public Map<String, String> getCanonicalizedExtensionHeaders() {
+    return canonicalizedExtensionHeaders;
+  }
+
+  public URI getCanonicalizedResource() {
+    return canonicalizedResource;
+  }
+
+  public static final class Builder {
+
+    private final HttpMethod httpVerb;
+    private String contentMd5;
+    private String contentType;
+    private final long expiration;
+    private Map<String, String> canonicalizedExtensionHeaders;
+    private final URI canonicalizedResource;
+
+    /**
+     * Constructs builder.
+     * 
+     * @param httpVerb  the HTTP method
+     * @param expiration  the EPOX expiration date
+     * @param canonicalizedResource  the resource URI
+     * @throws IllegalArgumentException if required field is not provided.
+     */
+    public Builder(HttpMethod httpVerb, long expiration, URI canonicalizedResource) {
+      checkArgument(httpVerb != null, "Required HTTP method");
+      checkArgument(canonicalizedResource != null, "Required canonicalized resource");
+
+      this.httpVerb = httpVerb;
+      this.expiration = expiration;
+      this.canonicalizedResource = canonicalizedResource;
+    }
+
+    public Builder(SignatureInfo signatureInfo) {
+      this.httpVerb = signatureInfo.httpVerb;
+      this.contentMd5 = signatureInfo.contentMd5;
+      this.contentType = signatureInfo.contentType;
+      this.expiration = signatureInfo.expiration;
+      this.canonicalizedExtensionHeaders = signatureInfo.canonicalizedExtensionHeaders;
+      this.canonicalizedResource = signatureInfo.canonicalizedResource;
+    }
+
+    public Builder setContentMd5(String contentMd5) {
+      this.contentMd5 = contentMd5;
+
+      return this;
+    }
+
+    public Builder setContentType(String contentType) {
+      this.contentType = contentType;
+
+      return this;
+    }
+
+    public Builder setCanonicalizedExtensionHeaders(
+        Map<String, String> canonicalizedExtensionHeaders) {
+      this.canonicalizedExtensionHeaders = canonicalizedExtensionHeaders;
+
+      return this;
+    }
+
+    /**
+     * Creates an {@code SignatureInfo} object from this builder.
+     */
+    public SignatureInfo build() {
+      return new SignatureInfo(this);
```

Comment:
Can there be a check to ensure that required fields (verb, resource, expiration) are non-null in the build() method?

Reply:
Done.

Moved null check validations from constructor to build method. Build pattern uses constructor parameters for required attributes, it is the reason why null check validations was there. It is true that the validations were lost in the constructor per copy, but thinking in "fail fast principle"... build() method is not the best place to perform that validations.

Since expiration is a primitive type, it does not required null check, instead I added a gte validation.